src/units/attack_type.hpp

```diff
@@ -83,6 +85,14 @@ class attack_type : public std::enable_shared_from_this<attack_type>
 	std::string weapon_specials() const;
 	std::string weapon_specials_value(const std::set<std::string> checking_tags) const;
 
+	const std::string& attack_alignment() const { return attack_alignment_; }
+	/** Returns alignment specified by 'attack_alignment' If empty and have unit's alignment returns the unit's alignment.
+	 */
+	unit_alignments::type alignment_in_attack() const;
+	/** Returns alignment specified by 'attack_alignment' for filtering.
+	 */
+	std::string effective_attack_alignment() const { return unit_alignments::get_string(alignment_in_attack()); }
```

Comment:
Why are 3 functions necessary to get the alignment when unit gets along with 1? IMO there should be a single function `alignment` that returns `unit_alignments::type` just like for units.

Reply:
This is because the alignment in [attack] must replace the unit alignment when it is correctly entered, otherwise it is the latter that is exercised, and this must be done in a function where the self_ variable is entered, this is why alignment_ in attack_tpe is a string variable. For the filtering check, it is the alignment actually used (if invalid alignment returns unit alignment) that is converted back to a string.

Finally, in the lua, self_ cannot be entered and it is therefore the alignment_ variable that is directly returned. So three functions are necessary

Reply:
I understand that you use all those functions but that does not seem reasonable and quite confusing to figure out from the names which function does what.

Internally you can use `alignment_` if you need the original string and otherwise you call `alignment()` which gives you the actual alignment. If you need the string then get the string from there.

If you really must have 3 different functions that do nearly the same thing they need to have _very_ good names to be able to tell what is what. But since unit gets along fine without that I don't see the point.

Reply:
@soliton- names changed, have you suggestion if these name are not convenient?

Reply:
Since you use `effective_alignment_str` you should also use `alignment_str`. I suppose that's clear enough then.

Reply:
Actually while the names are fine it's still not clear why `alignment_str()` is a useful public API. It seems to be the opposite it invites to hand out possibly invalid alignment strings.

Reply:
fixed with https://github.com/wesnoth/wesnoth/pull/8851/commits/b0fbac1393394babd181efc61f1c7748f96b2d61 commit

Reply:
@soliton- lokk at https://github.com/wesnoth/wesnoth/pull/8851/commits/b0fbac1393394babd181efc61f1c7748f96b2d61